File: packages/train/src/game/possibleMoves.ts
import { BIN_INDEXES, Board, EMPTY, Move, PLACE_INDEXES, Position } from './types';
import {
    getAt,
    getCardStackLength,
    getLastCards,
    getStackPossibleLength,
    isCompatible,
    isCompatibleBin,
} from './utils';
import { range } from '../utils';

const getCompatibleColumnIndexes = (
    card: string,
    lastCards: string[],
    excludeColumnIndex: number,
) => {
    const indexes: number[] = [];

    lastCards.forEach((lastCard, columnIndex) => {
        if (isCompatible(lastCard, card) && columnIndex !== excludeColumnIndex) {
            indexes.push(columnIndex);
        }
    });

    return indexes;
};

export const possibleMoves = (board: Board): Move[] => {
    const [lastCards, lastIndexes] = getLastCards(board);
    const maxStackLength = getStackPossibleLength(board);
    const maxEmptyColumnStackLength = getStackPossibleLength(board, true);
    const moves: Move[] = [];

    const getToLayoutMoves = (card: string, from: Position, stackLength: number) => {
        const column = from[1];
        const indexes = getCompatibleColumnIndexes(card, lastCards, column);

        return indexes
            .map((columnIndex) => {
                const toLevel = lastIndexes[columnIndex];
                return [toLevel === 1 ? 1 : toLevel + 1, columnIndex];
            })
            .filter(
                (to) =>
                    (to[0] === 1 && stackLength <= maxEmptyColumnStackLength) ||
                    (to[0] > 1 && stackLength <= maxStackLength),
            )
            .map((to) => [from, to] as Move);
    };
    const getToBinMoves = (card: string, from: Position) =>
        BIN_INDEXES.filter((columnIndex) =>
            isCompatibleBin(getAt(board, [0, columnIndex]), card),
        ).map((columnIndex) => [from, [0, columnIndex]] as Move);
    const getToPlaceMoves = (from: Position) =>
        PLACE_INDEXES.filter(
            (columnIndex) => getAt(board, [0, columnIndex]) === EMPTY,
        ).map((columnIndex) => [from, [0, columnIndex]] as Move);

    // from layout
    for (const column of range(board[0].length)) {
        for (let level = 1; level < board.length; level++) {
            const from = [level, column] as Position;
            const card = getAt(board, from);
            const stackLength = getCardStackLength(board, from);
            if (stackLength > 0) {
                // to layout
                const toLayoutMoves = getToLayoutMoves(card, from, stackLength);
                moves.push(...toLayoutMoves);
            }
            if (stackLength === 1) {
                // to place
                const toPlaceMoves = getToPlaceMoves(from);
                moves.push(...toPlaceMoves);

                // to bin
                const toBinMoves = getToBinMoves(card, from);
                moves.push(...toBinMoves);
            }
        }
    }

    // from place
    PLACE_INDEXES.forEach((columnIndex) => {
        const from = [0, columnIndex] as Position;
        const card = getAt(board, from);

        if (card) {
            // to layout
            const toLayoutMoves = getToLayoutMoves(card, from, 0);
            moves.push(...toLayoutMoves);

            // to bin
            const toBinMoves = getToBinMoves(card, from);
            moves.push(...toBinMoves);
        }
    });

    return moves;
};

File: pythonFramework/game/possible_moves.py
from .constants import EMPTY, BIN_INDEXES, PLACE_INDEXES
from .utils import isCompatible, getLastCards, getStackPossibleLength, getAt, getCardStackLength


def getCompatibleColumnIndexes(
        card,
        lastCards,
        excludeColumnIndex
):
    indexes = list()

    for (columnIndex, lastCard) in enumerate(lastCards):
        if isCompatible(lastCard, card) and columnIndex != excludeColumnIndex:
            indexes.append(columnIndex)

    return indexes


def possibleMoves(board):
    [lastCards, lastIndexes] = getLastCards(board)
    maxStackLength = getStackPossibleLength(board)
    maxEmptyColumnStackLength = getStackPossibleLength(board, True)
    moves = list()

    def getToLayoutMoves(card, fromPosition, stackLength):
        column = fromPosition[1]
        indexes = getCompatibleColumnIndexes(card, lastCards, column)

        def map_func(columnIndex):
            toLevel = lastIndexes[columnIndex]
            return [1 if toLevel == 1 else toLevel + 1, columnIndex]

        def filter_func(to):
            return (to[0] == 1 and stackLength <= maxEmptyColumnStackLength) or (
                    to[0] > 1 and stackLength <= maxStackLength)

        def to_move(to):
            return [fromPosition, to]

        return list(map(to_move, filter(filter_func, map(map_func, indexes))))

    def getToBinMoves(card, fromPosition):
        def filter_func(columnIndex):
            return getAt(board, [0, columnIndex]) == card

        def to_move(columnIndex):
            return [fromPosition, [0, columnIndex]]

        return list(map(to_move, filter(filter_func, BIN_INDEXES)))

    def getToPlaceMoves(fromPosition):
        def filter_func(columnIndex):
            return getAt(board, [0, columnIndex]) == EMPTY

        def to_move(columnIndex):
            return [fromPosition, [0, columnIndex]]

        return list(map(to_move, filter(filter_func, PLACE_INDEXES)))

    # from layout
    for column in range(len(board[0])):
        for level in range(1, len(board)):
            fromPosition = [level, column]
            card = getAt(board, fromPosition)
            stackLength = getCardStackLength(board, fromPosition)

            if stackLength > 0:
                # to layout
                moves.extend(getToLayoutMoves(card, fromPosition, stackLength))

            if stackLength == 1:
                # to place
                moves.extend(getToPlaceMoves(fromPosition))
                # to bin
                moves.extend(getToBinMoves(card, fromPosition))

    # from place
    for columnIndex in PLACE_INDEXES:
        fromPosition = [0, columnIndex]
        card = getAt(board, fromPosition)

        if card != EMPTY:
            # to layout
            moves.extend(getToLayoutMoves(card, fromPosition, 0))
            # to bin
            moves.extend(getToBinMoves(card, fromPosition))

    return moves
